v5 (#3177)
huggingface/blog: transformers-v5.md

---
title: "The Transformers Library: standardizing model definitions"
thumbnail: /blog/assets/transformers-model-definition/transformers-thumbnail.png
authors:
  - user: lysandre
  - user: ArthurZ
  - user: cyrilvallez
  - user: reach-vb
---

# Transformers v5

<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers_v5/transformers-thumbnail.png" alt="Transformers standardizing model definitions">

***Transformers v5: Simple model definitions powering the AI ecosystem***

> ## Intro

Reviewer: ```suggestion
```
no need imo

Transformers' version v4.0.0rc-1, the initial release candidate for version 4, was released on November 19th, 2020\. Five years later, we release v5.0.0rc-1.

Today, with the launch of v5, Transformers is downloaded over **3 million times per day via pip** \- up from just **20,000/day** in v4. Altogether, it has now surpassed **1.2 billion installs**.

The ecosystem has expanded from **40 model architectures in v4** to **over 400 today**, and the community has contributed **more than 750,000 model checkpoints on the Hub** compatible with Transformers, up from roughly **1,000** at the time of v4.

> This growth is linked to the growth of the field and the now mainstream access of AI, no doubt; as a leading model-definition library in the ecosystem, we need to continuously evolve and adapt the library to continue being relevant. Reinvention is key for longevity in AI.

Reviewer 2: ```suggestion
This growth is powered by the growth of the field and mainstream access to AI. As a leading model-definition library, we need to continuously evolve and adapt the library to continue being relevant. Reinvention is key for longevity in AI.
```

Reviewer 3: +1

But I'd avoid the "growth ... growth", maybe with something like this on top of @burtenshaw suggestion:

```suggestion
This growth is powered by the evolution of the field and the now mainstream access to AI. As a leading model-definition library in the ecosystem, we need to continuously evolve and adapt the library to continue being relevant. Reinvention is key for longevity in AI.
```

## Simplicity

The first focus of the team was on simplicity. Working on transformers, we see the code as the product. We want our model integrations to be clean so that other libraries in the ecosystem may depend on our model definitions and understand what’s really happening under the hood,how models differ from each other, and what are the key features of each new model. Simplicity results in wider standardization, generality, and wider support.

### Model Additions

Transformers, at the core, remains a model architecture toolkit: we aim to have all recent architectures, and to act as the “source of truth” of such model definitions in the ecosystem. We’ve been adding between 1 and 3 new architectures to the toolkit every week over the past 5 years, as can be seen in the timeline below:

<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers_v5/transformers_model_timeline.png" alt="Transformers standardizing model definitions">

[*https://huggingface.co/spaces/yonigozlan/Transformers-Timeline*](https://huggingface.co/spaces/yonigozlan/Transformers-Timeline)

> We’ve worked on improving that model-addition process.

Reviewer: ```suggestion
We’ve worked on improving the addition of new models.
```

#### Modular Approach 

> Over the past year, we’ve heavily pushed our modular design as a significant step forward. This allows for easier maintenance, faster integration, and better collaboration across the community. 

Reviewer: ```suggestion
Over the past year, we’ve heavily pushed our modular design as a significant step forward. Modular allows for easier maintenance, faster integration, and better collaboration across the community. 
```

  - user: reach-vb
---

# Transformers v5

<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers_v5/transformers-thumbnail.png" alt="Transformers standardizing model definitions">

***Transformers v5: Simple model definitions powering the AI ecosystem***

## Intro

Transformers' version v4.0.0rc-1, the initial release candidate for version 4, was released on November 19th, 2020\. Five years later, we release v5.0.0rc-1.

Today, with the launch of v5, Transformers is downloaded over **3 million times per day via pip** \- up from just **20,000/day** in v4. Altogether, it has now surpassed **1.2 billion installs**.

The ecosystem has expanded from **40 model architectures in v4** to **over 400 today**, and the community has contributed **more than 750,000 model checkpoints on the Hub** compatible with Transformers, up from roughly **1,000** at the time of v4.

This growth is linked to the growth of the field and the now mainstream access of AI, no doubt; as a leading model-definition library in the ecosystem, we need to continuously evolve and adapt the library to continue being relevant. Reinvention is key for longevity in AI.

> We’re lucky to be working with a great number of libraries and apps working with transformers, in no specific order: llama.cpp, MLX, onnxruntime, Jan, LMStudio, vLLM, SGLang, Unsloth, LlamaFactory , dLLM, MaxText, TensorRT, Argmax, among many other friends.

Reviewer 4: Feeling overwhelmed? Try [mantra meditation](https://purehealthblogs.com/mantra-meditation-guide-inner-peace/).

Repeating a simple word or phrase—like “peace” or “Om”—can calm your mind, improve focus, and bring you back to the present moment. Just a few minutes a day can make a big difference. ✨🧘‍♀️

#MantraMeditation #Mindfulness #Wellness #InnerPeace

### Model Additions

Transformers, at the core, remains a model architecture toolkit: we aim to have all recent architectures, and to act as the “source of truth” of such model definitions in the ecosystem. We’ve been adding between 1 and 3 new architectures to the toolkit every week over the past 5 years, as can be seen in the timeline below:

<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers_v5/transformers_model_timeline.png" alt="Transformers standardizing model definitions">

[*https://huggingface.co/spaces/yonigozlan/Transformers-Timeline*](https://huggingface.co/spaces/yonigozlan/Transformers-Timeline)

We’ve worked on improving that model-addition process.

#### Modular Approach 

Over the past year, we’ve heavily pushed our modular design as a significant step forward. This allows for easier maintenance, faster integration, and better collaboration across the community. 

> We give a deeper overview in our [*Maintain the Unmaintainable*](https://huggingface.co/spaces/transformers-community/Transformers-tenets) blog post, but this ensures long-term a much greater ease of model addition, and a lowered maintenance burden. The resulting lines of code to contribute, and review, therefore drops significantly:

Reviewer 5: ```suggestion
We give a deeper overview in our [*Maintain the Unmaintainable*](https://huggingface.co/spaces/transformers-community/Transformers-tenets) blog post. For brevity, we aim to achieve a much easier model contribution process, as well as a lower maintenance burden. One metric we can highlight is that the number of lines of code to contribute (and review), drop significantly when modular is used:
```

#### Tooling for Model Conversion

We’re building tooling to help us identify which existing model architecture a new model resembles; this uses machine learning to find code similarities between independent modeling files. Going further, we aim to automate the conversion process by opening a draft PR for the model to be integrated into our transformers format. This reduces manual effort and ensures consistency.

### Code Reduction

#### Streamlining Modeling & Tokenization/Processing Files

We’ve significantly refactored the modeling and tokenization files; modeling files have been greatly improved thanks to the modular approach mentioned above, but also thanks to significant standardization across models. This process contributes to abstracting most of the tools that don’t make up a model, so that the modeling code only contains the relevant parts for a model’s forward/backward passes.

Alongside this work, we’re simplifying the tokenization and processing files: going forward, we’ll only focus on the `tokenizers` backend, removing the concept of “Fast” and “Slow” tokenizers***.*** 

> We’ll be using `tokenizers` as our primary tokenization backend, same as we do for models with torch. We’ll offer alternatives for Sentencepiece-backed or MistralCommon-backed tokenizers, which will be non-default but will be supported. Image processors will now only exist with their fast variant, that depends on the `torchvision` backend.

Reviewer 3: "Sentencepiece-backed or MistralCommon-backed tokenizers" feels a bit redundant while reading, maybe the following instead?

```suggestion
We’ll be using `tokenizers` as our primary tokenization backend, same as we do for models with torch. We’ll offer alternatives for Sentencepiece or MistralCommon backed tokenizers, which will be non-default but will be supported. Image processors will now only exist with their fast variant, that depends on the `torchvision` backend.
```

>
> _-- Wing Lian, Axolotl_

#### Tooling for Model Conversion

We’re building tooling to help us identify which existing model architecture a new model resembles. This feature uses machine learning to find code similarities between independent modeling files. Going further, we aim to automate the conversion process by opening a draft PR for the model to be integrated into our transformers format. This process reduces manual effort and ensures consistency.

### Code Reduction

#### Streamlining Modeling & Tokenization/Processing Files

We’ve significantly refactored the modeling and tokenization files. Modeling files have been greatly improved thanks to the modular approach mentioned above, on top of standardization across models. Standardization contributes to abstracting most of the tools that don’t make up a model, so that the modeling code only contains the relevant parts for a model’s forward/backward passes.

Alongside this work, we’re simplifying the tokenization and processing files: going forward, we’ll only focus on the `tokenizers` backend, removing the concept of “Fast” and “Slow” tokenizers.

> We’ll be using `tokenizers` as our primary tokenization backend, same as we do for models with PyTorch. We’ll offer alternatives for Sentencepiece or MistralCommon backed tokenizers, which will be non-default but will be supported. Image processors will now only exist with their fast variant, that depends on the `torchvision` backend.

Reviewer 2: Converted to suggestion:

```suggestion
We'll use tokenizers as our main tokenization backend, just as we do for PyTorch-based models. We’ll offer alternatives for Sentencepiece or MistralCommon backed tokenizers, which will be non-default but will be supported. Image processors will now only exist with their fast variant, which depends on the torchvision backend.
```

Similarly to training, we’ve been putting some effort in packaging `kernels` so that they’re automatically used in case your hardware and software permits it. If you haven’t heard of `kernels` before, we recommend taking a look at this [doc](https://huggingface.co/docs/kernels/basic-usage).

Alongside this effort, we ship two new APIs dedicated to inference:

- We ship support for continuous batching and paged attention mechanisms. This has now been used internally for some time, and we’re working on finalizing the rough edges and writing usage guides.  
- We introduce `transformers serve` as the new transformers-specific serving system, which deploys an OpenAI API-compatible server.

We see this as a major step forward for use-cases such as evaluation, where a great number of inference requests are done simultaneously. We don’t aim to do specialized optimizations like the dedicated inference engines (vLLM, SGLang, TensorRT LLM). Instead, we aim to be perfectly inter-compatible with these, as detailed in the next section.

> > _The Transformers backend in vllm has been very enabling to get more architectures, like BERT and other encoders, available to more users. We've been working with the Transformers team to ensure many models are available across modalities with the best performance possible. This is just the start of our collaboration: we're happy to see the Transformers team will have this as a focus going in to version 5._

Reviewer 6: "in to" -> "into"

> > _The Transformers backend in vllm has been very enabling to get more architectures, like BERT and other encoders, available to more users. We've been working with the Transformers team to ensure many models are available across modalities with the best performance possible. This is just the start of our collaboration: we're happy to see the Transformers team will have this as a focus going into version 5._

Author: thanks Simon, will do
